metadatastore/commands.py

```diff
@@ -1,34 +1,94 @@
                         unicode_literals)
+
 import six
 from functools import wraps
 from itertools import count
-from .odm_templates import (RunStart, BeamlineConfig, RunStop,
-                            EventDescriptor, Event, DataKey, ALIAS)
-from .document import Document
+
 import datetime
 import logging
-from metadatastore import conf
-from mongoengine import connect,  ReferenceField
-import mongoengine.connection
 
-import datetime
+import boltons.cacheutils
 import pytz
 
-import uuid
-from bson import ObjectId
+from bson.dbref import DBRef
+
+from mongoengine import connect
+import mongoengine.connection
+
+from . import conf
+from .odm_templates import (RunStart, RunStop, EventDescriptor, Event, DataKey,
+                            ALIAS)
+from . import doc
+
 
 logger = logging.getLogger(__name__)
 
 
-__all__ = ['insert_beamline_config', 'insert_run_start', 'insert_event',
-           'insert_run_stop', 'insert_event_descriptor', 'find_run_stops',
-           'find_beamline_configs', 'find_event_descriptors', 'find_last',
-           'find_events', 'find_run_starts', 'db_connect', 'db_disconnect',
-           'format_data_keys', 'format_events', 'reorganize_event']
+# process local caches of 'header' documents these are storing object indexed
+# on ObjectId because that is what the reference fields in mongo are
+# implemented as.   Should move to uids asap
+_RUNSTART_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
+_RUNSTOP_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
+_DESCRIPTOR_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
+
+# never drop these
+_RUNSTART_UID_to_OID_MAP = dict()
+_RUNSTOP_UID_to_OID_MAP = dict()
+_DESCRIPTOR_UID_to_OID_MAP = dict()
+
+
+class NoRunStop(Exception):
+    pass
+
+
+class NoEventDescriptors(Exception):
+    pass
+
+
+def doc_or_uid_to_uid(doc_or_uid):
+    """Given Document or uid return the uid
+
+    Parameters
+    ----------
+    doc_or_uid : dict or str
+        If str, then assume uid and pass through, if not, return
+        the 'uid' field
+
+    Returns
+    -------
+    uid : str
+        A string version of the uid of the given document
+
+    """
+    if not isinstance(doc_or_uid, six.string_types):
+        doc_or_uid = doc_or_uid['uid']
+    return doc_or_uid
+
+
+def clear_process_cache():
+    """Clear all local caches"""
+    _RUNSTART_CACHE_OID.clear()
+    _RUNSTOP_CACHE_OID.clear()
+    _DESCRIPTOR_CACHE_OID.clear()
+
+    _RUNSTART_UID_to_OID_MAP.clear()
+    _RUNSTOP_UID_to_OID_MAP.clear()
+    _DESCRIPTOR_UID_to_OID_MAP.clear()
 
 
 def _ensure_connection(func):
+    """Decorator to ensure that the DB connection is open
+
+    This uses the values in `conf.connection_config` to get connection
+    parameters.
+
+    .. warning
+
+       This has no effect if the connection already exists, simply changing
+       the values in `conf.connection_config` is not sufficient to change what
+       data base is being used.
```

Comment:
Do you need to also call `db_disconnect()` for mongo to connect to a different database after you the config info is changed in `conf.connection_config`?